File: 2015/day04.cpp
// https://adventofcode.com/2015/day/4

#include "../common/common.h"
#include "../common/rust.h"

#ifndef AOC_IMPORT_STD
#include <algorithm>
#include <array>
#include <atomic>
#include <charconv>
#include <limits>
#include <print>
#include <span>
#include <thread>
#include <vector>
#endif

/// How many numbers are hashed side by side
constexpr let LANES = aoc::constant::md5_lanes;

auto parse(String const& filename) -> String {
  return String{aoc::trim(aoc::read_file(filename))};
}

/// Whether the hash starts with the five or six zeroes the part asks for
template <bool Part2>
fn is_answer(aoc::Digest const& hash) -> bool {
  if ((hash[0] != 0) || (hash[1] != 0)) {
    return false;
  }
  if constexpr (Part2) {
    return hash[2] == 0;
  } else {
    return (hash[2] & 0xF0u) == 0;
  }
}

template <bool Part2>
fn solve_case(str key) -> u32 {
  let num_threads = aoc::num_worker_threads();
  auto result = std::atomic<u32>{std::numeric_limits<u32>::max()};

  // A thread takes a whole batch at a time,
  // so the numbers it hashes side by side are the ones next to each other
  let step = num_threads * static_cast<u32>(LANES);
  auto search = [&, key](u32 start) {
    auto buffers = std::array<std::array<char, 64>, LANES>{};
    for (auto& buffer : buffers) {
      stdr::copy(key, std::begin(buffer));
    }
    auto messages = std::array<str, LANES>{};
    auto hashes = std::array<aoc::Digest, LANES>{};
    for (auto n = start; n < result.load(std::memory_order_relaxed);
         n += step) {
      for (let lane : Range{0uz, LANES}) {
        auto& buffer = buffers[lane];
        let[end, _] =
            std::to_chars(std::begin(buffer) + key.size(), std::end(buffer),
                          n + static_cast<u32>(lane));
        messages[lane] = str{std::begin(buffer), end};
      }
      aoc::md5_many(messages, hashes);

      // The batch runs upwards, so its first answer is its smallest
      for (let lane : Range{0uz, LANES}) {
        if (is_answer<Part2>(hashes[lane])) {
          let answer = n + static_cast<u32>(lane);
          auto current = result.load(std::memory_order_relaxed);
          while ((answer < current) &&
                 !result.compare_exchange_weak(current, answer,
                                               std::memory_order_relaxed))
            ;
          return;
        }
      }
    }
  };

  {
    auto threads = std::vector<std::jthread>{};
    threads.reserve(num_threads);
    for (u32 i = 0; i < num_threads; ++i) {
      threads.emplace_back(search, i * static_cast<u32>(LANES));
    }
  } // all threads join here

  return result.load();
}

int main() {
  std::println("Part 1");
  let example = parse("day04.example");
  AOC_EXPECT_RESULT(609043, solve_case<false>(example));
  AOC_EXPECT_RESULT(1048970, solve_case<false>("pqrstuv"));
  let input = parse("day04.input");
  AOC_EXPECT_RESULT(254575, solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(6742839, solve_case<true>(example));
  AOC_EXPECT_RESULT(5714438, solve_case<true>("pqrstuv"));
  AOC_EXPECT_RESULT(1038736, solve_case<true>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2015/src/day04.rs
use aoc::md5::{self, Digest, md5_many};
use aoc::string;
use std::sync::atomic::{AtomicU32, Ordering};
use std::thread;

fn parse(filename: &str) -> String {
    aoc::file::read_string(filename).trim().to_string()
}

/// Whether the hash starts with the five or six zeroes the part asks for
fn is_answer<const PART2: bool>(hash: &Digest) -> bool {
    if (hash[0] != 0) || (hash[1] != 0) {
        return false;
    }
    return if PART2 {
        hash[2] == 0
    } else {
        (hash[2] & 0xF0) == 0
    };
}

fn solve_case<const PART2: bool>(key: &str) -> u32 {
    let num_threads = thread::available_parallelism().map_or(1, |n| n.get()) as u32;
    let result = AtomicU32::new(u32::MAX);

    // A thread takes a whole batch at a time,
    // so the numbers it hashes side by side are the ones next to each other
    let step = num_threads * md5::LANES as u32;
    thread::scope(|s| {
        for thread in 0..num_threads {
            let result = &result;
            s.spawn(move || {
                let key_bytes = key.as_bytes();
                let mut buffers = [[0u8; 64]; md5::LANES];
                let mut sizes = [0; md5::LANES];
                for buffer in &mut buffers {
                    buffer[..key_bytes.len()].copy_from_slice(key_bytes);
                }
                let mut hashes = [Digest::default(); md5::LANES];
                let mut n = thread * md5::LANES as u32;
                while n < result.load(Ordering::Relaxed) {
                    for (lane, buffer) in buffers.iter_mut().enumerate() {
                        sizes[lane] = key_bytes.len()
                            + string::write_u32(&mut buffer[key_bytes.len()..], n + lane as u32);
                    }
                    let inputs: [&[u8]; md5::LANES] =
                        std::array::from_fn(|lane| &buffers[lane][..sizes[lane]]);
                    md5_many(&inputs, &mut hashes);

                    // The batch runs upwards, so its first answer is its smallest
                    if let Some(lane) = hashes.iter().position(is_answer::<PART2>) {
                        let answer = n + lane as u32;
                        let mut current = result.load(Ordering::Relaxed);
                        while answer < current {
                            match result.compare_exchange_weak(
                                current,
                                answer,
                                Ordering::Relaxed,
                                Ordering::Relaxed,
                            ) {
                                Ok(_) => break,
                                Err(c) => current = c,
                            }
                        }
                        return;
                    }
                    n += step;
                }
            });
        }
    });

    result.load(Ordering::Relaxed)
}

fn main() {
    println!("Part 1");
    let example = parse("day04.example");
    aoc::expect_result!(609043, solve_case::<false>(&example));
    aoc::expect_result!(1048970, solve_case::<false>("pqrstuv"));
    let input = parse("day04.input");
    aoc::expect_result!(254575, solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!(6742839, solve_case::<true>(&example));
    aoc::expect_result!(5714438, solve_case::<true>("pqrstuv"));
    aoc::expect_result!(1038736, solve_case::<true>(&input));
}
